### ISSUE TITLE
Live-lock in evaluator

### ISSUE THREAD

**Issue by mdekstrand** [lenskit/lenskit#840]
In certain rare cases, we encounter the live lock described in [JAVASERVERFACES-2544](https://java.net/jira/browse/JAVASERVERFACES-2544) in the evaluator.

Example stack trace:

```
"pool-1-thread-40" #61 prio=5 os_prio=0 tid=0x00007f1654a76000 nid=0x15825 runnable [0x00007eeee41eb000]
   java.lang.Thread.State: RUNNABLE
        at java.util.WeakHashMap.transfer(WeakHashMap.java:528)
        at java.util.WeakHashMap.resize(WeakHashMap.java:493)
        at java.util.WeakHashMap.put(WeakHashMap.java:466)
        at org.grouplens.grapht.InjectionContainer.makeInstantiator(InjectionContainer.java:118)
        at org.grouplens.grapht.InjectionContainer$DepLookup.apply(InjectionContainer.java:163)
        at org.grouplens.grapht.InjectionContainer$DepLookup.apply(InjectionContainer.java:143)
        at com.google.common.collect.Maps$AsMapView.get(Maps.java:793)
        at org.grouplens.grapht.reflect.internal.ClassInstantiator.instantiate(ClassInstantiator.java:93)
        at org.grouplens.grapht.Instantiators$ProviderInstantiator.instantiate(Instantiators.java:148)
        at org.grouplens.grapht.Instantiators$MemoizingInstantiator.instantiate(Instantiators.java:184)
        - locked <0x00007ef3c33aacb0> (a org.grouplens.grapht.Instantiators$MemoizingInstantiator)
        at org.lenskit.inject.NodeInstantiator$DefaultImpl.instantiate(NodeInstantiator.java:76)
        at org.lenskit.eval.traintest.ComponentCache$CacheEntry.getObject(ComponentCache.java:207)
        - locked <0x00007ef3c33aad50> (a org.lenskit.eval.traintest.ComponentCache$CacheEntry)
        at org.lenskit.eval.traintest.ComponentCache.processNode(ComponentCache.java:132)
        at org.lenskit.inject.NodeProcessors.processNodes(NodeProcessors.java:89)
        at org.lenskit.eval.traintest.ExperimentJob.buildRecommender(ExperimentJob.java:219)
        at org.lenskit.eval.traintest.ExperimentJob.run(ExperimentJob.java:98)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```

**Comment by mdekstrand** [comment]
hey @kluver, we're still seeing this in 3.0, thought you'd like to know so you can track status.

**Comment by mdekstrand** [comment]
I believe this is now fixed, with grouplens/grapht#134 and the associated LensKit changes in #841.

ping @kluver.

**Comment by kluver** [comment]
Ill try to remember to take a test pass at this when I have the time. On the assumption that this fixes it, thanks for the hard work @mdekstrand
